### ISSUE TITLE
Scan cache keys do not include commit SHA1s when tracking a moving branch

### ISSUE THREAD

**Issue by tardyp** [oss-review-toolkit/ort#4562]
The scanner decides which hash key it uses for the cache very early in the process.

This works well when using git tags in analyser, but is not appropriate when using branches (with option):

ort.analyzer.allowDynamicVersions=true
ort.downloader.allowMovingRevisions=true

This feature combination is useful for continuous integration workflow to always analyse the entire dependency tree tracking git branches, but only rescan when new commit is done in a repository (and also to share the cache between dependents)

**Comment by mnonnenmacher** [comment]
The scan storages always store the resolved revision, there should be no issue with using a branch. Also, from the description I do not understand what exactly the issue is.

Please provide more information:
* What exactly do you mean by "scan cache key"?
* Which storage are you using?
* What does not work as expected?
* Can you provide steps to reproduce the issue?

**Comment by tardyp** [comment]
> What exactly do you mean by "scan cache key"?

In order to implement a cache, usually we use some kind of key/value store, where the key is a hash of the input data.

> Which storage are you using?

The default file storage

> What does not work as expected?

The cache key only uses whatever was found by the analyser, and if the analyser does not have the formal revision, it will reuse the same scan result for the same branch.

> Can you provide steps to reproduce the issue?

my usecase is described in more details here #4505 , including a reproduction scenario.
Basically I am using the analyser to scan an spdx file, and then use the scanner scan the projects pointed by the spdx file.
In the spdx file, we only specify the branch, and not the revision.

In my tests, I found that I could only scan once, because the second try, the results was then cached. the scanner did fail to find the actual revision sha, so I am pretty sure the hashing key of the cache didn't include it.

**Comment by tardyp** [comment]
Here is a complete reproduction scenario:

https://gist.github.com/tardyp/1ac90495020214759bcf81cf8680a89e


When I run this script, I get the following file created:

```
~/.ort/scanner/results/SpdxDocumentFile/unknown/proj2/main/scan-results.yml.xz
```

The cache key in my wording would be `SpdxDocumentFile/unknown/proj2/main`. 
This does not include the revision information only `main`, which is a moving revision.

In order to fix this for moving revision, I would expect we may need to query the Downloader to get the current revision that is associated with the 'main' branch, and then store the result in 

```
~/.ort/scanner/results/SpdxDocumentFile/unknown/proj2/main-a32ed3b2297c7f3781590ae3ab4c38721abae13c/scan-results.yml.xz
```

**Comment by tardyp** [comment]
@mnonnenmacher in #4505 you suggest that resolved_revision is used inside the scan-results.

I can verify that it is indeed the case, but the resolved_revision is not checked to match a moving branch revision.
Thus, if I relaunch the [script](https://gist.github.com/tardyp/1ac90495020214759bcf81cf8680a89e) after pushing a commit in proj2, then the scan cache is not invalidated  

```
---
id: "SpdxDocumentFile::proj2:main"
results:
- provenance:
    vcs_info:
      type: "Git"
      url: "https://github.com/tardyp/proj1.git"
      revision: "2d88505742f9f28e4fddc6f73cfc100fd5998570"
      path: ""
    resolved_revision: "2d88505742f9f28e4fddc6f73cfc100fd5998570"
  scanner:
    name: "ScanCode"
    version: "21.6.7"
    configuration: "--copyright --license --info --strip-root --timeout 300 --json-pp"
  summary:
    start_time: "2021-09-30T19:10:56.356115Z"
    end_time: "2021-09-30T19:10:59.748434Z"
    package_verification_code: "a28acc0a33148fb27b8bb9bc15e0df815a25fc88"
    licenses:
    - license: "Apache-2.0"
      location:
        path: "LICENSE"
        start_line: 1
        end_line: 201
    copyrights: []
- provenance:
    vcs_info:
      type: "Git"
      url: "https://github.com/tardyp/proj2.git"
      revision: "main"
      path: ""
    resolved_revision: "fa42ed60a1dd8658d7c82390726e28d4481b977b"
  scanner:
    name: "ScanCode"
    version: "21.6.7"
    configuration: "--copyright --license --info --strip-root --timeout 300 --json-pp"
  summary:
    start_time: "2021-10-06T13:07:57.874772Z"
    end_time: "2021-10-06T13:08:01.944497Z"
    package_verification_code: "558c7e2dc237a05275e951a760a1b40abcf1b732"
    licenses:
    - license: "GPL-2.0-only"
      location:
        path: "LICENSE"
        start_line: 1
        end_line: 339
    copyrights:
    - statement: "Copyright (c) 1989, 1991 Free Software Foundation, Inc."
      location:
        path: "LICENSE"
        start_line: 4
        end_line: 6
    - statement: "copyrighted by the Free Software Foundation"
      location:
        path: "LICENSE"
        start_line: 251
        end_line: 256
```

**Comment by mnonnenmacher** [comment]
You are right, the root cause is that here the resolved revision is not used:
https://github.com/oss-review-toolkit/ort/blob/abd41579bd1e882870350dbcc7db781e22ec5c56/model/src/main/kotlin/Provenance.kt#L80

I will have to discuss this with the other core devs first before we can come up with a solution.

**Comment by sschuberth** [comment]
> I will have to discuss this with the other core devs first before we can come up with a solution.

Did we actually do this @mnonnenmacher? I cannot recall anymore 😆

**Comment by mnonnenmacher** [comment]
Yes, sorry for not commenting earlier. The outcome was that this cannot be fixed with the current scanner implementation, but it will work with the currently experimental scanner.

@tardyp You could give this a try with the `--experimental-scanners` option of the `scan` command.

**Comment by sschuberth** [comment]
> it will work with the currently experimental scanner.

Great, then let's close this once the "experimental scanner" becomes the stable default by middle of September 2022, also see https://github.com/oss-review-toolkit/ort/pull/5674.

**Comment by sschuberth** [comment]
@mnonnenmacher are we good to close this now that https://github.com/oss-review-toolkit/ort/pull/5799 is merged?

**Comment by mnonnenmacher** [comment]
Yes, moving revisions are now always resolved and rescanned if they have changed.
